File: easy/linked-list/python/19-remove-nth-node-from-end-of-list.py
from typing import List, Optional
from collections import defaultdict
# Definition for singly-linked list.


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def __str__(self) -> str:
        f = ""
        x = self
        while x is not None:
            f += f"{x.val}->"
            x = x.next
        return f


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        if head is None:
            return None
        nodes = {}
        ctr = 0
        while head is not None:
            nodes[ctr] = head
            head = head.next
            ctr += 1

        if n > ctr:
            return nodes[0]

        if n == ctr:
            return nodes[0].next

        # If we are deleting the last node, n + 1 node will not exist
        if n == 1:
            nodes[ctr - n - 1].next = None
            return nodes[0]

        nodes[ctr - n].val = nodes[ctr - n + 1].val
        nodes[ctr - n].next = nodes[ctr - n + 1].next
        return nodes[0]


# Regular remove
l = ListNode(1, ListNode(2, ListNode(3, ListNode(4, (ListNode(5))))))
print(Solution().removeNthFromEnd(l, 2))

# Remove only 1 item
l = ListNode(1)
print(Solution().removeNthFromEnd(l, 1))

# Empty list
l = None
print(Solution().removeNthFromEnd(l, 1))

# Remove last item
l = ListNode(1, ListNode(2))
print(Solution().removeNthFromEnd(l, 1))

# Remove first item
l = ListNode(1, ListNode(2, ListNode(3)))
print(Solution().removeNthFromEnd(l, 3))

File: easy/linked-list/cpp/19-remove-nth-node-from-end-of-list.cpp
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

#include "../../../stdc++.h"

using std::unordered_map;

class Solution
{
public:
    ListNode *removeNthFromEnd(ListNode *head, int n)
    {
        unordered_map<int, ListNode *> map;
        int counter = 0;
        while (head != nullptr)
        {
            map[counter] = head;
            head = head->next;
            counter++;
        }
        counter--;

        // Remove 1st item
        if (n == counter + 1)
        {
            return map[0]->next;
        }

        // Remove last item
        if (n == 0)
        {
            // There is only 1 item in the list
            if (counter == 0)
            {
                return nullptr;
            }

            // Make next item of second last item null
            map[counter - 1]->next = nullptr;
            return map[0];
        }

        map[counter - n]->next = map[counter - n + 2];
        return map[0];
    }
};
void print(ListNode *head)
{
    while (head != nullptr)
    {
        std::cout << head->val << std::endl;
        head = head->next;
    }
}
int main()
{
    Solution a;
    ListNode *ptr;
    // ListNode *ptr = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
    // a.removeNthFromEnd(ptr, 3);
    // print(ptr);

    // std::cout << "==========" << std::endl;

    // ptr = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
    // a.removeNthFromEnd(ptr, 2);
    // print(ptr);

    // std::cout << "==========" << std::endl;

    // ptr = new ListNode(10);
    // a.removeNthFromEnd(ptr, 1);

    // std::cout << "==========" << std::endl;

    ptr = new ListNode(1, new ListNode(2));
    ListNode *res = a.removeNthFromEnd(ptr, 2);
    print(res);
}
